utilities/fault_injection_fs.cc

```diff
@@ -479,12 +480,54 @@ size_t TestFSRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
     return target_->GetUniqueId(id, max_size);
   }
 }
+
+void FaultInjectionTestFS::AddUnsyncedToRead(const std::string& fname,
+                                             size_t offset, size_t n,
+                                             Slice* result, char* scratch) {
+  // Should be checked prior
+  assert(result->size() < n);
+  size_t offset_after = offset + result->size();
+
+  MutexLock l(&mutex_);
+  auto it = db_file_state_.find(fname);
+  if (it != db_file_state_.end()) {
+    auto& st = it->second;
+    if (st.pos_ > static_cast<ssize_t>(offset_after)) {
+      size_t remaining_requested = n - result->size();
+      size_t to_copy = std::min(remaining_requested,
+                                static_cast<size_t>(st.pos_) - offset_after);
+      size_t buffer_offset =
+          offset_after -
+          static_cast<size_t>(std::max(st.pos_at_last_sync_, ssize_t{0}));
+      // Data might have been dropped from buffer
+      if (st.buffer_.size() > buffer_offset) {
+        to_copy = std::min(to_copy, st.buffer_.size() - buffer_offset);
+        if (result->data() != scratch) {
+          // NOTE: might overlap
+          assert(false);
```

Comment:
I actually don't understand - why do we have result->data() != scratch?

Reply:
Read() doesn't have to use the beginning of scratch, or scratch at all, for returning data. In practice this only arises with mmap reads, which is only used for FSRandomAccessFile. Since I'm not implementing those cases, I've made this a TODO item.